aptos/contract/aux/sources/stake.move

```diff
@@ -0,0 +1,1571 @@
+        pool.acc_reward_per_share = pool.acc_reward_per_share + duration_reward * REWARD_PER_SHARE_MUL / (total_stake as u128);
+        pool.reward_remaining = pool.reward_remaining - (duration_reward as u64);
+        // the real balance of the reward coin should always be >= the virtual remaining balance
+        assert!(coin::value(&pool.reward) >= pool.reward_remaining, E_INTERNAL_ERROR);
+
+        pool.last_update_time = now;
+
+    }
+
+    /// Deposit stake coin to the incentive pool to start earning rewards.
+    /// All pending rewards will be transferred to `sender`.
+    public entry fun deposit<S, R>(sender: &signer, id: u64, amount: u64) acquires Pools, UserPositions {
+        assert!(amount > 0, E_INVALID_DEPOSIT_AMOUNT);
+        assert!(exists<Pools<S, R>>(@aux), E_INCENTIVE_POOL_NOT_FOUND);
+        let pools = borrow_global_mut<Pools<S, R>>(@aux);
+        assert!(table::contains(&mut pools.pools, id), E_INCENTIVE_POOL_NOT_FOUND);
+        let pool = table::borrow_mut(&mut pools.pools, id);
+        let now = timestamp::now_microseconds();
+        update_pool(pool, now);
+
+        // Deposit staked coin
+        let stake = coin::withdraw<S>(sender, amount);
+        coin::merge(&mut pool.stake, stake);
+
+        // Update UserPosition
+        let sender_addr = signer::address_of(sender);
+        if (!exists<UserPositions<S, R>>(sender_addr)) {
+            move_to(sender, UserPositions<S, R> {
+                positions: table::new()
+            })
+        };
+        let positions = borrow_global_mut<UserPositions<S, R>>(sender_addr);
+        let exists = table::contains(&positions.positions, id);
+        let deposit_event = if (!exists) {
+            let reward_debt = (amount as u128) * pool.acc_reward_per_share / REWARD_PER_SHARE_MUL;
+            table::add(&mut positions.positions, id, UserPosition<S, R> {
+                amount_staked: amount,
+                reward_debt
+            });
+            DepositEvent<S, R> {
+                pool_id: id,
+                user: sender_addr,
+                deposit_amount: amount,
+                reward_amount: 0,
+                reward_debt,
+                amount_staked: amount,
+                total_amount_staked: coin::value(&pool.stake),
+                reward_remaining: pool.reward_remaining,
+                acc_reward_per_share: pool.acc_reward_per_share
+            }
+        } else {
+            let user_pos = table::borrow_mut(&mut positions.positions, id);
+
+            // Distribute pending rewards
+            let new_reward_num = (user_pos.amount_staked as u128) * pool.acc_reward_per_share;
+            if (new_reward_num < pool.undistributed_reward) {
+                pool.undistributed_reward = pool.undistributed_reward - new_reward_num;
+            } else {
+                pool.undistributed_reward = 0;
+            };
+            let pending_reward = new_reward_num / REWARD_PER_SHARE_MUL - user_pos.reward_debt;
```

Comment:
can we make sure pending reward is round down first, then minus the reward debt?

Reply:
Updated to use Nikhil's suggestion of storing `last_acc_reward_per_share` instead, which avoids the rounding issue.